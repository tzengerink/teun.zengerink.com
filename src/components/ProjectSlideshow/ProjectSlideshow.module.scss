@import '../../styles/index';

.slideshow {
  position: absolute;
}

.slide {
  display: none;
  text-align: center;
  font-size: 1em;
  line-height: 0;
}

.previous,
.next {
  position: absolute;
  top: 0;
  bottom: 0;
}

.previous {
  left: 0;
  right: 50%;
  cursor: w-resize;
}

.next {
  left: 50%;
  right: 0;
  cursor: e-resize;
}

.active {
  display: inline-block;
}

.statement {
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;

  .text {
    @include transition(0.3s);
    @include transform(translateY(20px));

    position: fixed;
    left: 0;
    bottom: 0;
    right: 0;
    opacity: 0;
    visibility: hidden;
    padding: $margin;
    margin-bottom: $margin;
    font-size: 0.9em;
    background-color: $background-color-transparent;
    z-index: 1;
  }

  h2 {
    position: fixed;
    bottom: $margin;
    left: $margin;
    margin-bottom: 0;
    text-transform: uppercase;
    font-size: 0.7em;
    z-index: 2;
    cursor: default;

    &:hover {
      display: block;
    }

    &:hover + .text {
      @include transform(translateY(0px));

      display: block;
      opacity: 1;
      visibility: visible;
      cursor: pointer;
    }
  }

  h3 {
    margin-top: 0;
  }
}

@include from-tablet-portrait {
  .slideshow {
    @include transform(translate(-50%, -50%));

    top: 50%;
    left: 50%;
  }

  .statement {
    display: block;

    .text {
      left: $margin;
      bottom: $margin;
      padding: $margin $margin $margin 0;
      margin-bottom: 0;
      width: 500px;
    }

    h2 {
      right: auto;
      font-size: 0.8em;

      &:hover + .text {
        display: block;
        cursor: pointer;
      }
    }
  }
}
