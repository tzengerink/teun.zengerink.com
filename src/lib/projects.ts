import fs from 'fs'
import path from 'path'
import slugify from 'slugify'
import yaml from 'js-yaml'

export enum Width {
  Mobile = 768,
  Desktop = 1280,
}

interface ConfigCaption {
  key: string
  caption: string
}

interface ConfigItem {
  title: string
  statement?: string
  isArchived?: boolean
  captions?: ConfigCaption[]
}

export interface Export {
  width: Width
  url: string
}

export interface Photo {
  key: string
  caption?: string
  exports: Export[]
}
export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
  isArchived?: boolean
}

export const MAX_WIDTHS = [Width.Mobile, Width.Desktop]

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), 'public/photos', slug, `${Width.Mobile}w`)
  const fileNames = fs.readdirSync(photosDirectory)

  return fileNames.map((filename) => {
    const key = path.basename(filename, '.jpg')
    const caption = captions?.find((c) => c.key == key)

    return {
      key,
      caption: caption?.caption ?? '',
      exports: MAX_WIDTHS.map((width) => ({ width, url: `/photos/${slug}/${width}w/${filename}` })),
    }
  })
}

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true })

    return {
      slug,
      title: project.title,
      statement: project.statement ?? '',
      isArchived: project.isArchived ?? false,
      photos: generatePhotos(slug, project.captions),
    }
  })
}

export const getProjects = async (): Promise<Project[]> => {
  let projects: Project[] = []

  try {
    const file = path.join(process.cwd(), 'src', 'config.yml')
    const config = yaml.load(fs.readFileSync(file, 'utf-8'))
    projects = await generateProjects(config)
  } catch (e) {
    console.error(e)
  }

  return projects
}
