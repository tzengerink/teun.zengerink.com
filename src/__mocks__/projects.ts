import { Project } from '../lib/projects'

const projects: Project[] = [
  {
    title: 'My First Project',
    slug: 'my-first-project',
    statement: 'The *statement* for the _project_',
    photos: [
      {
        key: '01',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/01.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/02.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/02.jpg` },
        ],
      },
      {
        key: '03',
        exports: [
          { width: 768, url: `/photos/my-first-project/768w/03.jpg` },
          { width: 1280, url: `/photos/my-first-project/1280w/03.jpg` },
        ],
      },
    ],
  },
  {
    title: 'Project without a statement',
    slug: 'no-statement',
    photos: [
      {
        key: '01',
        caption: 'This is _image 1_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/01.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        caption: 'This is _image 2_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/02.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/02.jpg` },
        ],
      },
      {
        key: '03',
        caption: 'This is _image 3_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/03.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/03.jpg` },
        ],
      },
      {
        key: '04',
        caption: 'This is _image 4_',
        exports: [
          { width: 768, url: `/photos/no-statement/768w/04.jpg` },
          { width: 1280, url: `/photos/no-statement/1280w/04.jpg` },
        ],
      },
    ],
  },
  {
    title: 'Archived project',
    slug: 'archived-project',
    isArchived: true,
    photos: [
      {
        key: '01',
        exports: [
          { width: 768, url: `/photos/archived-project/768w/01.jpg` },
          { width: 1280, url: `/photos/archived-project/1280w/01.jpg` },
        ],
      },
      {
        key: '02',
        exports: [
          { width: 768, url: `/photos/archived-project/768w/02.jpg` },
          { width: 1280, url: `/photos/archived-project/1280w/02.jpg` },
        ],
      },
    ],
  },
]

export default projects
