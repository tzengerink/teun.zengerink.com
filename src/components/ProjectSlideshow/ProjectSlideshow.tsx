import classNames from 'classnames'
import React, { useEffect } from 'react'
import { Project } from '../../lib/types'
import { useProject } from '../../lib/useProject'
import Photo from '../Photo/Photo'
import Statement from '../Statement/Statement'

enum Key {
  LEFT = 'ArrowLeft',
  RIGHT = 'ArrowRight',
}

interface ProjectSlideshowProps {
  project: Project
}

const ProjectSlideshow: React.FC<ProjectSlideshowProps> = ({ project }) => {
  let touchStartX: number
  const { activeKey, previous, next } = useProject(project)

  useEffect(() => {
    const keyUpHandler = ({ key }: KeyboardEvent) => {
      switch (key) {
        case Key.LEFT:
          return previous()
        case Key.RIGHT:
          return next()
      }
    }

    const ON = 'keyup'
    window.addEventListener(ON, keyUpHandler)
    return () => window.removeEventListener(ON, keyUpHandler)
  })

  const touchStartHandler = ({ touches }: React.TouchEvent) => {
    touchStartX = touches[0].clientX
  }

  const touchMoveHandler = ({ touches }: React.TouchEvent) => {
    if (!touchStartX) return
    const difference = touchStartX - touches[0].clientX
    if (difference < 0) previous()
    if (difference > 0) next()
    touchStartX = null
  }

  return (
    <div className={classNames('relative', 'h-[calc(100vh-150px)]', 'md:h-[100vh]')}>
      <div
        className={classNames('absolute', 'top-1/2', 'translate-y-[-50%]', 'sm:left-1/2', 'sm:translate-x-[-50%]')}
        onTouchStart={(e) => touchStartHandler(e)}
        onTouchMove={(e) => touchMoveHandler(e)}
      >
        <div
          data-testid="left-side"
          className={classNames('absolute', 'top-0', 'bottom-0', 'left-0', 'right-1/2', 'cursor-w-resize')}
          onClick={() => previous()}
        />
        <div
          data-testid="right-side"
          className={classNames('absolute', 'top-0', 'bottom-0', 'right-0', 'left-1/2', 'cursor-e-resize')}
          onClick={() => next()}
        />
        {project?.photos.map((photo) => (
          <Photo
            key={`${project.slug}--slide-${photo.key}`}
            className={classNames('max-h-[calc(100vh-150px)]', 'sm:max-h-[100vh]', 'text-center', {
              block: photo.key === activeKey,
              hidden: photo.key !== activeKey,
            })}
            photo={photo}
            alt={`${project?.title} - ${photo.key}`}
          />
        ))}
      </div>
      {project?.statement ? <Statement title={project.title} statement={project.statement} /> : null}
    </div>
  )
}

export default ProjectSlideshow
