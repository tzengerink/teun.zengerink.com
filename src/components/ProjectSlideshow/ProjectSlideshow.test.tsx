import { fireEvent, render } from '@testing-library/react'
import React from 'react'
import projects from '../../__mocks__/projects'
import ProjectSlideshow from './ProjectSlideshow'

const mockProject = projects[0]
const mockNext = jest.fn()
const mockPrevious = jest.fn()

jest.mock('../../lib/useProject', () => ({
  useProject: () => ({
    activeKey: mockProject.photos[2].key,
    next: () => mockNext(),
    previous: () => mockPrevious(),
  }),
}))

describe('ProjectSlideshow', () => {
  afterEach(() => jest.clearAllMocks())

  const renderComponent = (project = mockProject, index = '01') =>
    render(<ProjectSlideshow project={project} index={index} />)

  it('renders correctly for projects with a statement', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('renders correctly for projects with captions', () => {
    const { container } = renderComponent(projects[1])
    expect(container).toMatchSnapshot()
  })

  describe('Keyboard navigation', () => {
    const expectKeyupToCall = (key, mock) => {
      renderComponent()
      expect(mock).not.toHaveBeenCalled()
      window.dispatchEvent(new KeyboardEvent('keyup', { key }))
      expect(mock).toHaveBeenCalled()
    }

    it('triggers next when right arrow key is pressed', () => {
      expectKeyupToCall('ArrowRight', mockNext)
    })

    it('triggers previous when left arrow key is pressed', () => {
      expectKeyupToCall('ArrowLeft', mockPrevious)
    })
  })

  describe('Mouse navigation', () => {
    const expectClickToCall = (element, mock) => {
      expect(mock).not.toHaveBeenCalled()
      fireEvent.click(element)
      expect(mock).toHaveBeenCalled()
    }

    it('triggers next when clicking right side of the slideshow', () => {
      const { getByTestId } = renderComponent()
      const rightSide = getByTestId('right-side')
      expectClickToCall(rightSide, mockNext)
    })

    it('triggers previous when clicking left side of the slideshow', () => {
      const { getByTestId } = renderComponent()
      const leftSide = getByTestId('left-side')
      expectClickToCall(leftSide, mockPrevious)
    })
  })

  describe('Touch navigation', () => {
    const expectSwipeToCall = (element, direction: 'left' | 'right', mock) => {
      const clientX = 10
      expect(mock).not.toHaveBeenCalled()
      fireEvent.touchMove(element, { touches: [{ clientX }] })
      expect(mock).not.toHaveBeenCalled()
      fireEvent.touchStart(element, { touches: [{ clientX }] })
      expect(mock).not.toHaveBeenCalled()
      const newClientX = direction === 'left' ? clientX - 5 : clientX + 5
      fireEvent.touchMove(element, { touches: [{ clientX: newClientX }] })
      expect(mock).toHaveBeenCalled()
    }

    it('triggers next when swiping a photo to the left', () => {
      const { getByAltText } = renderComponent()
      const photo = getByAltText(`${mockProject.title} - ${mockProject.photos[0].key}`)
      expectSwipeToCall(photo, 'left', mockNext)
    })

    it('triggers previous when swiping a photo to the right', () => {
      const { getByAltText } = renderComponent()
      const photo = getByAltText(`${mockProject.title} - ${mockProject.photos[0].key}`)
      expectSwipeToCall(photo, 'right', mockPrevious)
    })
  })
})
