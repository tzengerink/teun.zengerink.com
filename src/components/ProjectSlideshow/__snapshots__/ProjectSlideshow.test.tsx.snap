// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectSlideshow renders correctly for projects with a statement 1`] = `
<div>
  <div
    class="relative h-[calc(100vh-150px)] md:h-[100vh]"
  >
    <div
      class="absolute top-1/2 translate-y-[-50%] sm:left-1/2 sm:translate-x-[-50%]"
    >
      <div
        class="absolute top-0 bottom-0 left-0 right-1/2 cursor-w-resize"
        data-testid="left-side"
      />
      <div
        class="absolute top-0 bottom-0 right-0 left-1/2 cursor-e-resize"
        data-testid="right-side"
      />
      <img
        alt="My First Project - 01"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center block"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/my-first-project/1280w/01.jpg"
        srcset="/photos/my-first-project/768w/01.jpg 768w, /photos/my-first-project/1280w/01.jpg 1280w"
      />
      <img
        alt="My First Project - 02"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/my-first-project/1280w/02.jpg"
        srcset="/photos/my-first-project/768w/02.jpg 768w, /photos/my-first-project/1280w/02.jpg 1280w"
      />
      <img
        alt="My First Project - 03"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/my-first-project/1280w/03.jpg"
        srcset="/photos/my-first-project/768w/03.jpg 768w, /photos/my-first-project/1280w/03.jpg 1280w"
      />
    </div>
    <div
      class="group text-grey/80 select-none"
    >
      <h2
        class="fixed left-md bottom-md z-20 text-xs uppercase cursor-help"
      >
        Statement
      </h2>
      <div
        class="fixed left-0 bottom-md z-10 right-0 mb-md p-md bg-transparant-white invisible opacity-0 ease-linear duration-100 translate-y-[50px] group-hover:visible group-hover:translate-y-0 group-hover:opacity-100 md:w-[500px]"
      >
        <h3
          class="mb-sm text-sm"
        >
          My First Project
        </h3>
        <div
          class="space-y-sm text-xs text-justify"
        >
          <p>
            The 
            <em>
              statement
            </em>
             for the 
            <em>
              project
            </em>
          </p>
          

        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProjectSlideshow renders correctly for projects with captions 1`] = `
<div>
  <div
    class="relative h-[calc(100vh-150px)] md:h-[100vh]"
  >
    <div
      class="absolute top-1/2 translate-y-[-50%] sm:left-1/2 sm:translate-x-[-50%]"
    >
      <div
        class="absolute top-0 bottom-0 left-0 right-1/2 cursor-w-resize"
        data-testid="left-side"
      />
      <div
        class="absolute top-0 bottom-0 right-0 left-1/2 cursor-e-resize"
        data-testid="right-side"
      />
      <img
        alt="Project without a statement - 01"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center block"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/no-statement/1280w/01.jpg"
        srcset="/photos/no-statement/768w/01.jpg 768w, /photos/no-statement/1280w/01.jpg 1280w"
      />
      <img
        alt="Project without a statement - 02"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/no-statement/1280w/02.jpg"
        srcset="/photos/no-statement/768w/02.jpg 768w, /photos/no-statement/1280w/02.jpg 1280w"
      />
      <img
        alt="Project without a statement - 03"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/no-statement/1280w/03.jpg"
        srcset="/photos/no-statement/768w/03.jpg 768w, /photos/no-statement/1280w/03.jpg 1280w"
      />
      <img
        alt="Project without a statement - 04"
        class="w-auto h-auto md:max-w-[70vw] md:max-h-[90vh] max-h-[calc(100vh-150px)] sm:max-h-[100vh] text-center hidden"
        sizes="(max-width: 767px) 768px, 1280px"
        src="/photos/no-statement/1280w/04.jpg"
        srcset="/photos/no-statement/768w/04.jpg 768w, /photos/no-statement/1280w/04.jpg 1280w"
      />
    </div>
  </div>
</div>
`;
